toolchain/sem_ir/value_stores.h

```diff
@@ -21,7 +22,11 @@ class InstStore {
   // instruction block. Check::Context::AddInst or InstBlockStack::AddInst
   // should usually be used instead, to add the instruction to the current
   // block.
-  auto AddInNoBlock(Inst inst) -> InstId { return values_.Add(inst); }
+  auto AddInNoBlock(Parse::NodeId parse_node, Inst inst) -> InstId {
+    // TODO: CARBON_CHECK the parse node kind against the instruction kind.
+    parse_nodes_.push_back(parse_node);
+    return values_.Add(inst);
+  }
```

Comment:
I think the loss of type safety here is something we should address in the short term. I think this can be improved by making `AddInNoBlock` a template:

```suggestion
  template <typename InstT>
  auto AddInNoBlock(typename decltype(InstT::Kind)::TypedNode parse_node, InstT inst) -> InstId {
    parse_nodes_.push_back(parse_node);
    return values_.Add(inst);
  }
```

We might also need an overload for the case where `InstT` is exactly `Inst` that checks the parse node kind at runtime.

```c++
  template <typename InstT>
  auto AddInNoBlock(Parse::NodeId parse_node, InstT inst)
      -> std::enable_if_t<std::is_same_v<InstT, Inst>, InstId> {
    // TODO: CARBON_CHECK the parse node kind against the instruction kind.
    parse_nodes_.push_back(parse_node);
    return values_.Add(inst);
  }
```

We'll presumably want to do the same thing everywhere we're taking a `NodeId` and an `Inst` in order to not regress type safety, such as in `Context::AddInst`, and in `PendingBlock`. Might be worth adding a node-and-inst pair type that carries an inst and a suitable kind of node for it, to factor out the checks?

Reply:
Most of the calls are indirect. Switched to ParseNodeAndInst to handle validation.

Reply:
Noting the type-checking changes are significant, I figured it's better to go through another review round than just merge. Changing the call semantics in order to get compile-time type checking changes all the call sites, so I didn't think it's worth just merging -- I don't think it materially changes the churn.